Add oc adm compare command

AlinaSecret opened the pull request:
This commit introduces a new command that intelligently diffs a known valid reference configuration against a set of specific cluster configuration resources. The command matches each resource in the cluster configuration with a template in the reference configuration, injects the templated resource with cluster resource parameters, and presents users with a detailed diff highlighting differences between each cluster resource and its corresponding injected template. This command supports in validating adaptations of configurations against the original reference, identifying and addressing potential issues.

liornoy commented on pkg/cli/admin/admin.go:
```diff
@@ -3,6 +3,8 @@ package admin
 import (
 	"fmt"
 
+	"github.com/openshift/oc/pkg/cli/admin/compare"
```
Could be cleaner to move this together with the "github.com/openshift/oc/pkg/cli/admin/*" imports group in lines 17-41

liornoy commented on pkg/cli/admin/compare/compare.go:
```diff
@@ -0,0 +1,394 @@
+
+const (
+	ReferenceFileName        = "metadata.yaml"
+	noReffDirectoryWasPassed = "\"Reference directory is required\""
+	reffDirNotExistsError    = "\"Reference directory doesnt exist\""
```
Looking at other "oc" error messages I think you'll have to lowercase + remove the `\"`

liornoy commented on pkg/cli/admin/compare/groups/groups.go:
```diff
@@ -0,0 +1,61 @@
+		groups = append(groups, make(map[G][]E))
+	}
+	for _, element := range elements {
+		parsedElement, err := parsing(element)
```
suggesting "parse" as the function name here
save as having "get" function vs "getting"

AlinaSecret replied:
Definitely!

liornoy commented on pkg/cli/admin/compare/groups/groups.go:
```diff
@@ -0,0 +1,61 @@
+				errs = append(errs, fmt.Errorf("error while finding group hash for %v. Error: %v", parsedElement, groupErr))
+				break
+			}
+			_, ok := groups[i][groupName]
+			if !ok {
+				groups[i][groupName] = []E{}
+			}
```
it's common to one-line statements like those to:
if _, ok := groups[i][groupName]; !ok {
...

AlinaSecret replied:
sure thing

liornoy commented on pkg/cli/admin/compare/corelator.go:
```diff
@@ -0,0 +1,343 @@
+	for i, mapping := range mappings {
+		res := groups.GetWithMoreThen(mapping, 1)
+		if res != nil {
+			klog.Warningf("More Then one template with same %s. These templates wont be used for"+
```
```suggestion
			klog.Warningf("More then one template with same %s. These templates wont be used for"+
```

AlinaSecret replied:
good eye

imiller0 commented on pkg/cli/admin/compare/compare.go:
```diff
@@ -0,0 +1,394 @@
+		Compare a known valid reference configuration and a set of specific cluster configuration CRs.
+		
+		The reference configuration consists of Resource templates. 
+		Resource Templates are files that contain Resource definitions and are formatted as Go templates.
+		The compare command will match each Resource in the cluster configuration to a template in the reference 
+		configuration. Then, the templated Resource will be injected with the cluster Resource parameters. 
+		For each cluster Resource, a diff between the Resource and its matching injected template will be presented
+    	to the user.
```
Minor nits
```suggestion
		Resource Templates are files that contain Resource definitions and with fixed and optional content. Optional content is represented as Go templates.
		The compare command will match each Resource in the cluster configuration to a Resource Template in the reference 
		configuration. Then, the templated Resource will be injected with the cluster Resource parameters. 
		For each cluster Resource, a diff between the Resource and its matching injected template will be presented
    	to the user.
```

AlinaSecret replied:
A lot better!

imiller0 commented on pkg/cli/admin/compare/compare.go:
```diff
@@ -0,0 +1,394 @@
+		resulting in a report to the user in case one of them is missing.
+		
+		Each Resource definition should be in its own template file. 
+		The usage of all Go built-in functions is supported along with the functions in the Sprig library.
```
We should clarify here that the input to the Go template is the "input cluster configuration" in order to allow expected user variable content to be synchronized between cluster CR and reference CR prior to the diff.

AlinaSecret replied:
Added it :)

imiller0 commented on pkg/cli/admin/compare/compare.go:
```diff
@@ -0,0 +1,394 @@
+		All templates should always be valid YAML after template execution, even when injecting an empty mapping.
+		Before using functions that can fail for nil values, always check that the value exists.
+		
+		Each Resource definition should be in a separate template file, and the use of all Go built-in functions, 
+        along with those in the Sprig library, is supported. It is important that all templates remain valid YAML 
+		after execution, even in scenarios where an empty mapping is injected. 
+		Prior to using functions that may encounter nil values, it is critical to verify the existence of the value.
```
This is a repeat of the prior paragraph :)

AlinaSecret replied:
Ops right, my mistake

imiller0 commented on pkg/cli/admin/compare/compare.go:
```diff
@@ -0,0 +1,394 @@
+		after execution, even in scenarios where an empty mapping is injected. 
+		Prior to using functions that may encounter nil values, it is critical to verify the existence of the value.
+
+		It's possible to pass a user config that contains an option to specify manual matches between cluster resources
+		and Resource templates. The matches can be added to the config as pairs of 
+		apiVersion-kind-namespace-name: <Template File Name>.
```
```suggestion
		In some instances the correlation of user config to a reference template is ambiguous. In those cases it's possible to pass an option to explicitly identify the correlation. The matches can be added to the config as pairs of 
		apiVersion-kind-namespace-name: <Template File Name>.
```

imiller0 replied:
Maybe also clarify how to use this option for Cluster scoped CRs which do not have a namespace.

AlinaSecret replied:
added an explanation

imiller0 commented on pkg/cli/admin/compare/compare.go:
```diff
@@ -0,0 +1,394 @@
+			// `oc compare` propagates the error code from
+			//`kubectl diff` that propagates the error code from
+			// diff or `KUBECTL_EXTERNAL_DIFF`. Also, we
+			// don't want to print an error if diff returns
```
I agree we don't want to print an error if it returns "1" (indicating diffs were found), but having a non-zero return code would allow the command to be used in scripts/automation contexts for auto validation.

AlinaSecret replied:
totally agree with you, view lines 141-147.  It behaves as you describe. I thought that the behavior of this part (lines 117- 147) is exactly the same as in oc diff, therefore i copied it from there , including the comment.

imiller0 commented on pkg/cli/admin/compare/compare.go:
```diff
@@ -0,0 +1,394 @@
+		"Number of objects to process in parallel when diffing against the live version. Larger number = faster,"+
+			" but more memory, I/O and CPU over that shorter period of time.")
+	kcmdutil.AddFilenameOptionFlags(cmd, &options.CRs, "contains the configuration to diff")
+	cmd.Flags().StringVarP(&options.userConfigFileName, "user-config", "c", "", "Path to the user config file")
```
Maybe refer to this as the "diff config" file? calling it a user config sounds similar the input cluster config.

AlinaSecret replied:
sounds good didn't think of that

imiller0 commented on pkg/cli/admin/compare/compare.go:
```diff
@@ -0,0 +1,394 @@
+	return o.setLiveSearchTypes(f)
+}
+
+func (o *Options) setCorelator() error {
```
Method level comment would be helpful to explain how multiple correlators are used and the significance of ordering them.
The method name implies that it is setting a single correlator but I believe this is actually setting up the full set of correlators which will implement the specific match ordering rules that the tool works through. Specifically, user defined correlations followed by the successively less specific tuples defined in fieldGroups.

AlinaSecret replied:
I added More documentation to this function, also changed the name so it wont be misleading. 
I transferred the fieldGroups variable to this function i think its a lot more clear if it appears in this context.

imiller0 commented on pkg/cli/admin/compare/compare.go:
```diff
@@ -0,0 +1,394 @@
+			return !ok
+		})
+		sort.Strings(notSupportedTypes)
+		klog.Warningf("Reference Contains Templates With Types (kind) Not Supported By Cluster: %s", strings.Join(notSupportedTypes, ", "))
```
Ok for this to be a warning. This may be a common case where optional day-2 operators are not installed and thus their CRDs are not defined on the cluster.

AlinaSecret replied:
great, good to know

imiller0 commented on pkg/cli/admin/compare/corelator.go:
```diff
@@ -0,0 +1,343 @@
+
+func apiKindNamespaceName(r *unstructured.Unstructured) string {
+	if r.GetNamespace() == "" {
+		return strings.Join([]string{r.GetAPIVersion(), r.GetKind(), r.GetName()}, fieldSeparator)
+	}
+	return strings.Join([]string{r.GetAPIVersion(), r.GetKind(), r.GetNamespace(), r.GetName()}, fieldSeparator)
```
The field separator (currently "-") is a valid character in names (or namespaces). There may be the possibility that a cluster scoped CR with name "abc-def" would incorrectly match a namespaced CR with namespace "abc" and name "def". Choosing a character for the separator that is invalid in the fields would ensure this can't happen.

AlinaSecret replied:
changed it to _

imiller0 commented on pkg/cli/admin/compare/corelator.go:
```diff
@@ -0,0 +1,343 @@
+		if res != nil {
+			klog.Warningf("More Then one template with same %s. These templates wont be used for"+
+				" corelation. To use them use diffrent corelator (manual matching) or remove one of them from the"+
+				" reffernce  Template names are: %s", getFields(fieldGroups[i]), getTemplatesName(res))
```
```suggestion
				" refernce  Template names are: %s", getFields(fieldGroups[i]), getTemplatesName(res))
```

AlinaSecret replied:
fixed :)

imiller0 commented on pkg/cli/admin/compare/corelator.go:
```diff
@@ -0,0 +1,343 @@
+// (with groups with less `indexed fields`) until it finds a distinct match, in case it doesn't, MultipleMatches error
+// will be returned.
+// Templates will be only indexed by a group of fields only if all fields in group are not templated.
+type GroupCorelator struct {
```
When someone outside of our team reviews this it would be helpful to describe this in terms of what user's will be doing with this tool. The group correlator (if I'm following correctly) will be used to support "close matches" when there is not an exact match in which case this allows matching by api/kind/namespace or api/kind or just kind. Is that right?

imiller0 commented on pkg/cli/admin/compare/testdata/AllRequiredTemplatesExistAndThereAreNoDiffs/reference/cm.yaml:
```diff
@@ -0,0 +1,7 @@
+  labels:
+    k8s-app: kubernetes-dashboard
+  name: kubernetes-dashboard-settings
+  namespace: kubernetes-dashboard
```
All files should have a trailing newline to avoid this warning in github

AlinaSecret replied:
fixed in all files

imiller0 commented on pkg/cli/admin/compare/compare.go:
```diff
@@ -0,0 +1,394 @@
+		# Compare a known valid reference configuration with a live cluster and with a user config:
+		oc adm compare -r ./reference -c ./user_config
+	`)
+	fieldGroups = [][][]string{
+		{{"apiVersion"}, {"metadata", "name"}, {"metadata", "namespace"}, {"kind"}},
+		{{"apiVersion"}, {"metadata", "name"}, {"kind"}},
+		{{"metadata", "namespace"}, {"kind"}},
+		{{"kind"}},
+	}
```
I believe these go with the group correlator and define the "search order" when an exact match is not found. It woud be good to document that here.
It might be helpful to have a note here on how references with templated names (or namespaces) are covered under this search logic. Specifically how various combinations of templated name and/or namespace will get correlated.

AlinaSecret replied:
i moved it to the the setupCorelators function in compare.go, there it is also passed to the GroupCorelator. 
added there a comment explaining the "match  attempt order"  and examples for a combination of how a:
- templated namespace and fixed apiVersion, name, and kind, 
 - templated name  and fixed  apiVersion, name, and kind
 will get correlated. didn't add all the combinations, it think its clear enough, and don't want to add to many text.  would like to hear what you think? 
 
 also we should decide to maybe add more groups to improve correlation, for example we can add :
 ({"metadata", "name"}, {"kind"}}. that will give better correlation in case there is a template with fixed name and but but templated  apiVersion and name.

imiller0 reviewed (commented):
Really nice work so far! Thanks for posting this review :) 
I think some additional comments/documentation on how the logic addresses the specific use case will help (I left comments in various places) especially for the correlation logic and some of our corner cases.

Great work. I look forward to taking it for a drive.

AlinaSecret committed:
Add oc adm compare command

This commit introduces a new command that intelligently diffs a known valid
reference configuration against a set of specific cluster configuration
resources. The command matches each resource in the cluster configuration with
a template in the reference configuration, injects the templated resource with
cluster resource parameters, and presents users with a detailed diff
highlighting differences between each cluster resource and its corresponding
injected template. This command supports in validating adaptations of
configurations against the original reference, identifying and addressing
potential issues.

Signed-off-by: Alina Sudakov <asudakov@redhat.com>

AlinaSecret committed:
Add tests for oc adm compare

This commit adds tests for the `oc adm compare` command. The test cases include comparing CRs from both a live cluster and a local directory. These tests run `oc adm compare` with various outputs and verify that the expected stdout and stderr are obtained. A `--update` flag has been introduced to enable easy test writing and modification. The tests execute the command with a concurrency value greater than 1, enabling race detection when running the tests with the `-race` flag.

Signed-off-by: Alina Sudakov <asudakov@redhat.com>

pixelsoccupied commented on pkg/cli/admin/compare/compare_test.go:
```diff
@@ -0,0 +1,311 @@
+		require.NoError(t, cmd.Flags().Set("filename", resourcesDir))
+		require.NoError(t, cmd.Flags().Set("recursive", "true"))
+		break
+	case Live:
+		discoveryResources, resources := getResources(t, resourcesDir)
+		updateTestDiscoveryClient(tf, discoveryResources)
+		setClient(t, resources, tf)
+		break
```
Thought `envtest` library is a good one to flag...not sure if it's worth it here but something for internal testing env down the road... https://book.kubebuilder.io/reference/envtest.html

AlinaSecret committed:
Add option to pull reference config from a URL

In this commit, the option to pass a URL with the -r flag containing a raw
version of the reference config was introduced. To support this change, HTTPFS
was introduced, implementing a file system interface for HTTP servers
returning raw file content as part of the HTTP response body. Additionally,
the template.ParseFiles function was replaced with the template.ParseFS function,
allowing retrieval of the reference from a local file system or an HTTP file
system. Additional tests were added to support this case.

Furthermore, the change included calling parseFS each time with one file
instead of multiple files, enabling the reporting of all template parsing
errors to the user in one run. This also 
fixed problems with templates
included in subdirectories, which were reported missing in the summary even if
they were not.

Signed-off-by: Alina Sudakov <asudakov@redhat.com>

openshift-merge-bot[bot] committed:
Merge pull request #1704 from ingvagabund/extend-tools-image-with-sosreport

WRKLDS-1103: tools: extend the image with sosreport

AlinaSecret committed:
Improve live cluster search by using GroupVersionKind 

This update modifies how resources are found in live cluster mode. Previously,
we searched for resources based only on their  kind, rather
than their GroupVersionKind. Now, 
the code is adjusted to search for each resource kind found in the reference
CR across all API group-versions where that kind exists. To enable this
change, a list of group-versions is generated for each supported kind. This
resolves issues where the same type was returned by multiple API resources,
causing the visitor to miss some resources of that type.

Signed-off-by: Alina Sudakov <asudakov@redhat.com>
